### DISCUSSION TITLE
Possible regression on SIGQUIT handling in v5.5.0rc2

### DISCUSSION THREAD

**Discussion by Lotram** [celery/celery#9433] [Q&A]
Hi,

I've been playing with signals and celery 5.4.0 to properly log killed tasks (by my kubernetes cluster)
Since v5.5.0 is going to be released really soon, I tried to run the same tests on it. 
While SIGTERM and SIGKILL seem to behave the same as before, It does not seem to be the case for SIGQUIT.
I'm using:
```
celery=5.5.0rc2
billiard==4.2.1
kombu==5.5.0rc2
```
My `celery_app.py` file:
```
from time import sleep

from celery import Celery

app = Celery("proj", broker="amqp://guest@localhost//")

# app.conf.update(worker_soft_shutdown_timeout=1.2)


@app.task(bind=True)
def sleep_for(self, seconds):
    print("start to sleep")
    sleep(seconds)
    print("finished sleeping")

# Plus a print on every worker and task signal
```
When I run `celery -A celery_app worker -l info -c 2` and run `sleep_for.delay(100)`, then kill the celery worker with the QUIT signal (only this process, not the 2 children, maybe that's the issue ?), I get:
```
# first SIGQUIT
worker: Hitting Ctrl+C again will terminate all running tasks!
[2024-11-25 10:51:11,516: WARNING/ForkPoolWorker-1] worker_process_shutdown sender: None

# second SIGQUIT 
Waiting gracefully for cold shutdown to complete...

worker: Cold shutdown (MainProcess)
worker_shutting_down signal sender: celery@hostname

# third SIGQUIT
[2024-11-25 10:52:04,045: WARNING/ForkPoolWorker-2] task_postrun: 570f4237-18d5-4767-9925-62c71a28be27, retval: None
[2024-11-25 10:52:04,045: WARNING/ForkPoolWorker-2] worker_process_shutdown sender: None
```
If I don't send the third SIGQUIT, the worker waits for the task to be finished. This was not the case in celery 5.4.0, the task was terminated right away. 

If I use the `worker_soft_shutdown_timeout` setting, I get a similar behaviour:
```
# first SIGQUIT
worker: Hitting Ctrl+C again will terminate all running tasks!
[2024-11-25 10:55:20,559: WARNING/MainProcess] Initiating Soft Shutdown, terminating in 1.2 seconds
[2024-11-25 10:55:21,760: WARNING/ForkPoolWorker-1] worker_process_shutdown sender: None

# second SIGQUIT
Waiting gracefully for cold shutdown to complete...

worker: Cold shutdown (MainProcess)
worker_shutting_down signal sender: celery@hostname

# third SIGQUIT
[2024-11-25 10:55:53,385: WARNING/ForkPoolWorker-2] task_postrun: 844c6694-47f9-4f9b-93b2-43b22b7b6004, retval: None
[2024-11-25 10:55:53,385: WARNING/ForkPoolWorker-2] worker_process_shutdown sender: None
```
This not what the [doc is saying](https://docs.celeryq.dev/en/latest/userguide/workers.html#stopping-the-worker) about the SIGQUIT signal. Should I open an issue, or am I missing something ?

**Comment by Nusnus** [comment]
Interesting report!

Try adding `task_acks_late=True`, and it will work:
`app = Celery("proj", broker="amqp://guest@localhost//", task_acks_late=True)`
![CleanShot 2024-11-25 at 16 07 26@2x](https://github.com/user-attachments/assets/71878faa-e49a-4ff6-9be1-6dac66390a4c)


That being said, this behavior is indeed a change from 5.4, where `task_acks_late=True` wasn’t needed for SIGQUIT to terminate the tasks immediately, nor does it mention it specifically in the docs which can be misleading.

It’s been a while since I worked on it. Let me review the changes further and decide how to move forward.

Thank you! @Lotram 

P.S
Next time feel free to open an official bug report, but don’t waste your time now, this is enough as I could easily reproduce the behavior based on your description.

**Reply by Nusnus** [reply]
This is caused due to this line:
https://github.com/celery/celery/blob/1b35d1d5966614ce36af75808ee21b0d2db6745d/celery/worker/consumer/consumer.py#L740

Which is called during the first SIGQUIT, but without `task_acks_late=True` skips cancelling the `sleep_for` task because it is already acknowledged.

**Reply by Lotram** [reply]
Ok thanks for the answer. Why do we want to tie the shutdown behavior to tasks being acked late or not ? 

On that subject, I've just discovered the [task_reject_on_worker_lost](https://docs.celeryq.dev/en/stable/userguide/configuration.html#task-reject-on-worker-lost) setting. 
If I'm not mistaken, the message is indeed acked when the worker process receives a sigkill, but not when the worker itself receives it, which is usually the case when using containers. Is this the expected behavior ?

**Reply by Nusnus** [reply]
> Ok thanks for the answer. Why do we want to tie the shutdown behavior to tasks being acked late or not ?

To allow a graceful shutdown, graceful meaning respecting the acknowledgement configuration and not just dropping everything and closing.

> On that subject, I've just discovered the [task_reject_on_worker_lost](https://docs.celeryq.dev/en/stable/userguide/configuration.html#task-reject-on-worker-lost) setting. If I'm not mistaken, the message is indeed acked when the worker process receives a sigkill, but not when the worker itself receives it, which is usually the case when using containers. Is this the expected behavior ?

See how it is used here; take note that it is also tied to the `acks_late` configuration for the same graceful reasons.
https://github.com/celery/celery/blob/69904ee59a9e6dfacad10aa9af7001f1ca36ac1c/celery/worker/request.py#L602-L618

**Comment by Nusnus** [comment]
Fixed in #9461
